_posts/2022-04-09-how-eex-turns-your-template-into-html.md

```diff
@@ -0,0 +1,288 @@
+---
+layout: post
+title: "How EEx Turns Your Template Into HTML"
+date: 2022-04-09 09:00:00 -04:00
+categories: post
+permalink: /:title/
+
+---
+
+EEx is a templating language and module built into the [Elixir](https://hexdocs.pm/eex/EEx.html) standard library, often used by web servers to render dynamic content for a web application. The same concept exists in other language ecosystems, ERB for Ruby, Blade for Laravel/PHP, Pug/Jade for JavaScript.
+
+A couple features of EEx that make it stand out are:
+
+- Templates can be compiled into functions ahead of time, avoiding needing to read from the file system at runtime. 
+- EEx has the concept of an "Engine" allowing you to compile an existing template into something different than what the builtin engine, [`EEx.SmartEngine`](https://hexdocs.pm/eex/EEx.SmartEngine.html), does. This is what [Phoenix LiveView](https://hexdocs.pm/phoenix_live_view/Phoenix.LiveView.html) does.
+
+Recently, [José Valim tweeted me some advice on how to use the EEx Engine](https://twitter.com/josevalim/status/1511348149323014155?s=20&t=sFyaicmJApbs5yQO-n_CIA) in my library [Temple](https://github.com/mhanberg/temple) (an alternative HTML library, that uses an Elixir DSL). At first I was very confused at José's suggestion, I don't want to write EEx at all! I want Temple to go from DSL straight to `iolist` or `%LiveView.Rendered{}` structs.
+
+Luckily, my coworker [Marlus Saraiva](https://twitter.com/MarlusSaraiva) messaged me and gave a bit more context on Josés suggestion, and suddenly it all clicked!. (Hopefully at the end of this post, you'll understand how José's suggestion was makes perfect sense and was extremely clear.)
+
+This conversation led me to ask myself, how does EEx work anyway? I decided to source dive EEx, LiveView, [Surface](https://surface-ui.org/), and [HEEx](https://hexdocs.pm/phoenix_live_view/Phoenix.LiveView.Helpers.html#sigil_H/2-syntax) and will share with you what I learned!
+
+## The Source Code
+
+Technically EEx is just a templating syntax that can handle any plaintext format, it doesn't matter if its HTML, JavaScript, or even Elixir (this is how the Phoenix generators work).
+
+EEx templates allow you do execute arbitrary Elixir code inside of special tags and inject the runtime result into the return value. You can use single expressions or even expressions that take blocks. The `EEx.SmartEngine` also allows you to ergonomically access data passed to the template with the `@` syntax.
+
+```eex
+<!-- single line expression -->
+<span><%= @user.name %></span>
+
+<!-- block expression -->
+<ul>
+  <%= for user <- @users do %>
+    <li><%= user.name %></li>
+  <% end %>
+</ul>
+```
+
+## The Compiler
+
+The public API for compiling an EEx file or string are the `EEx.compile_file/2` and `EEx.compile_string/2` functions. Internally, they will call into the `EEx.Compiler` module.
```

Comment:
I think I wold say is instead of are because API is singular even though you mention multiple functions that make up the API in the same sentence.